### PR TITLE
Refactor serial port usage

### LINKED ISSUE #91: Reduce coupling on libserialport
I'm building m8c for Android downstream, with a hope of upstreaming it at some point if there is interest. Unrooted Androids do not seem to have direct access to serial ports, instead you get the USB device you want to use in Java side (dealing with permissions etc.),open a connection, claim the CDC/ACM interfaces and pass the file descriptor to the native code where you can communicate with the device.
Since libserialport doesn't deal with file descriptors, something like libusb needs to be used instead.

So my proposal is to:
- Move dependencies to libserialport.h to one file
- Create more abstract functions to deal with serial init, read and write. Something that doesn't have dependencies on libserialport.h
- Use something like `#ifdef LIBSERIALPORT` in the .c files to turn libserialport dependency on and off.

I'm not a C developer in any capacity so ideas are welcome.

This is something I'm going to do anyway so it's not a request for somebody else to do it, more like a check to see if I should attempt to upstream it as well.

### PR DESCRIPTION
**Pull request by v3rm0n**
- Move all references to libserialport to serial.c
- Move the serial port instance to serial.c and don't expose it
- Define USE_LIBSERIALPORT flag to turn libserialport dependent code on/off

Closes #91

### CODE REVIEW

**Comment by v3rm0n** [comment]
Actually I would like to change it so that libserialport is always the default and I'll add libusb implementation which can be turned on by specifying USE_LIBUSB

**Comment by v3rm0n** [comment]
I'm marking it as a draft for now beecause I want to improve it a bit. I don't like the way that the filehandle works currently and it's pretty hard to do other libusb related stuff (like audio). Will work on it a bit more.

**Comment by v3rm0n** [comment]
I think it looks better now.

File: main.c [hunk 1]
```diff
@@ -56,21 +49,20 @@ int main(int argc, char *argv[]) {
...
   static slip_handler_s slip;
-  struct sp_port *port = NULL;
 
   uint8_t prev_input = 0;
   uint8_t prev_note = 0;
   uint16_t zerobyte_packets = 0; // used to detect device disconnection
 
   signal(SIGINT, intHandler);
   signal(SIGTERM, intHandler);
+  signal(SIGQUIT, intHandler);
```
Comment (laamaa) [inline main.c:59]: This doesn't work under Windows, it might need to be wrapped in a "#ifdef SIGQUIT"

**Comment by v3rm0n** [comment]
Probably some parameters need a bit of tuning to get it to work optimally. I just used whatever started to work (like `#define serial_read_size 512`)

File: Makefile [hunk 2]
```diff
@@ -27,6 +27,10 @@ EXTENSION = .c
 m8c: $(OBJ)
 	$(CC) -o $@ $^ $(local_CFLAGS) $(INCLUDES)
 
+libusb: INCLUDES = $(shell pkg-config --libs sdl2 libusb)
+libusb: local_CFLAGS = $(CFLAGS) $(shell pkg-config --cflags sdl2 libusb) -Wall -O2 -pipe -I.
```
Comment (laamaa) [inline Makefile:31]: This probably needs -DUSE_LIBUSB somewhere to get the libusb code actually baked in :)

File: Makefile [hunk 3]
```diff
@@ -1,8 +1,8 @@
 #Set all your object files (the object files of all the .c files in your project, e.g. main.o my_sub_functions.o )
-OBJ = main.o serial.o slip.o command.o write.o render.o ini.o config.o input.o font.o fx_cube.o
+OBJ = main.o serial.o slip.o command.o render.o ini.o config.o input.o font.o fx_cube.o
```
Comment (laamaa) [inline Makefile:2]: We need to have "usb.o" here

File: Makefile [hunk 4]
```diff
@@ -27,6 +27,10 @@ EXTENSION = .c
 m8c: $(OBJ)
 	$(CC) -o $@ $^ $(local_CFLAGS) $(INCLUDES)
 
+libusb: INCLUDES = $(shell pkg-config --libs sdl2 libusb)
```
Comment (laamaa) [inline Makefile:30]: On my system pkg-config wanted libusb-1.0 as the parameter ... but I'm not sure if it's just my distro (Fedora 37)?

Comment (v3rm0n) [reply Makefile]: Yeah, same actually on a mac :)

**Comment by laamaa** [comment]
Thank you!! :)

**Pull request merged**